doc/developer/design/20231110_aws_connections.md

```diff
@@ -0,0 +1,164 @@
+      },
+      "Action": "sts:AssumeRole",
+      "Condition": {
+        "StringEquals": {
+          "sts:ExternalId": "<external_id>"
+        }
+      }
+    }
+  ]
+}
+```
+
+## Implementation
+We already have some [existing code](https://github.com/MaterializeInc/materialize/blob/v0.77.1/src/storage-types/src/connections/aws.rs), which can be re-used to
+handle the AssumeRole scenario.
+
+In the parser we already support `ROLE ARN` option which can be renamed to `ASSUME ROLE ARN`
+to better reflect the usage. We will need to add an `ASSUME ROLE SESSION NAME` option as well.
+
+The AWS Connection create statement with all the options will now look like:
+```sql
+CREATE CONNECTION <connector_name> TO AWS (
+  ACCESS KEY ID [[=] <value>], -- existing
+  SECRET ACCESS KEY [[=] SECRET <value>], -- existing
+  TOKEN [[=] SECRET <value>], -- existing
+  ASSUME ROLE ARN [[=] <value>], -- rename existing from ROLE ARN
+  ASSUME ROLE SESSION NAME [[=] <value>], -- new option
+  ENDPOINT [[=] <value>], -- existing
+  REGION [[=] <value>] -- existing
+);
+```
+
+Users should be able to provide either the `ACCESS KEY` options or the `ASSUME ROLE` options,
+but not both.
+
+#### External ID and principal
+We already have external ID prefix provided in the catalog state.
+The connection ID will be appended to the external ID prefix to get the complete
+External ID and then stored in the new `mz_aws_connections` table for that connection ID.
+
+For principal, we'll create one global role which all environmentd/clusterd-s can assume.
+This will add some good restrictions on who could assume the customers role and also
+allow for migration between environment clusters.
+
+Note: There's a hard limit of 1 hour on the session duration when [Role chaining in AWS](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_terms-and-concepts.html).
+This can be revisited for workarounds if we hit this issue.
+
+Other alternative solutions were:
+1. Using our root account `arn:aws:iam::<account_id>:root`. This would work but this is not
+secure enough since any user/role in this account can assume the customer's role.
+2. An IAM user `arn:aws:iam::<account_id>:user/<id>` created for the customer's environment similar
+to [Snowflake](https://docs.snowflake.com/en/user-guide/data-load-s3-config-storage-integration#step-4-retrieve-the-aws-iam-user-for-your-snowflake-account).
+This could work if we keep them in a secret and rotate regularly, this would give
+each customer a unique principal. But we'd have to rotate very frequently and we may also
+need to coordinate that role across all users of the credentials which seems like a lot of overhead.
+3. An IAM role `arn:aws:iam::<account_id>:role/<id>` we create for each customer environment. This
+will not scale, there's a maximum limit of [5000 roles](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html#reference_iam-quotas-entities)
+we can create per account.
+
+#### `VALIDATE CONNECTION`
```

Comment:
One thing to call out here: the default for the `VALIDATE` option for `CREATE CONNECTION ... TO AWS` needs to be `FALSE`. This is not the usual case, but it matches SSH connections. The reason is that, for both of these connection types, the user needs to do some additional setup that depends on the created connection. For SSH connections, the user has to authorize the connection's SSH public keys, which can only be acquired after the connection is created. Similarly, for AWS connections, the user has to write an IAM policy that authorizes the external ID for the connection, which again can only be acquired after the connection is created. 

Assuming you agree with the above, would be great to make a note in the design doc!

Reply:
That makes sense. And yes, until the user adds the trust policy on their end, validation would not work. I will add a note for this.

Reply:
Updated in the `CREATE CONNECTION` section above.